src/kixi/datastore/metadatastore/command/handler.clj

```diff
@@ -114,166 +120,175 @@
+
+(defmacro define-metadata-update-implementations
+  []
+  `(do
+     ~@(map define-metadata-update-implementation metadata-types->field-actions)))
+
+(define-metadata-update-implementations)
 
 (spec/def ::metadata-update
   (spec/multi-spec metadata-update :metadata-update))
 
 (defn invalid
   [cmd speccy data]
-  {:kixi.comms.event/key :kixi.datastore.metadatastore/update-rejected
-   :kixi.comms.event/version "1.0.0"
-   :kixi.comms.event/payload {:reason :invalid
+  {::ke/key ::kdm/update-rejected
+   ::ke/version "1.0.0"
+   ::ke/payload {:reason :invalid
                               :explanation (spec/explain-data speccy data)
                               :original {::ms/payload cmd}
-                              :kixi/user (:kixi.comms.command/user cmd)}})
+                              :kixi/user (::kc/user cmd)}})
+
+(defn semantically-invalid
+  [cmd explanation]
+  {::ke/key ::kdm/update-rejected
+   ::ke/version "1.0.0"
+   ::ke/payload {:reason :invalid
+                              :explanation explanation
+                              :original {::ms/payload cmd}
+                              :kixi/user (::kc/user cmd)}})
 
 (defn unauthorised
   [cmd id]
-  {:kixi.comms.event/key :kixi.datastore.metadatastore/update-rejected
-   :kixi.comms.event/version "1.0.0"
-   :kixi.comms.event/payload {:reason :unauthorised
+  {::ke/key ::kdm/update-rejected
+   ::ke/version "1.0.0"
+   ::ke/payload {:reason :unauthorised
                               ::ms/id id
-                              :kixi/user (:kixi.comms.command/user cmd)}})
+                              :kixi/user (::kc/user cmd)}})
 
 (defn updated
   [payload]
-  {:kixi.comms.event/key :kixi.datastore.file-metadata/updated
-   :kixi.comms.event/version "1.0.0"
-   :kixi.comms.event/payload payload})
+  {::ke/key ::kdfm/updated
+   ::ke/version "1.0.0"
+   ::ke/payload payload})
 
-(defn assoc-types
-  [md payload]
-  (merge payload
-         (select-keys md
-                      [::ms/type ::ms/bundle-type])))
-
-(defn valid-update
+(defn structurally-valid
   [metadatastore {:keys [::ms/id] :as payload}]
   (spec/valid? ::metadata-update 
                payload))
 
-(defn resolve-packed-ids
-  [metadatastore user-groups current updates]
-  (if (= "bundle"
-         (::ms/type updates))
-    (let [current-ids (set (::ms/packed-ids current))
-          updated-ids (set (::ms/packed-ids updates))
-          removed (clojure.set/difference current-ids updated-ids)
-          added (clojure.set/difference updated-ids current-ids)
-          removable (apply disj removed (unauthorised-ids metadatastore user-groups removed))
-          addable (apply disj added (unauthorised-ids metadatastore user-groups added))]
-      (assoc updates
-             ::ms/packed-ids
-             (->> current-ids
-                  (concat addable)
-                  (remove (set removable))
-                  vec)))
-    updates))
+(defn not-semantically-valid
```

Comment:
There is already a `semantically-invalid` function - can there be a better differentiation?